view containers integration

akosyakov opened the pull request:
New container contributed by the plugin system:
<img width="1792" alt="Screen Shot 2019-07-09 at 16 53 28" src="https://user-images.githubusercontent.com/3082655/60899049-1ab7d880-a26a-11e9-85da-e9faf84ec1d1.png">

Views contributed to the explorer container by the plugin system:
<img width="1792" alt="Screen Shot 2019-07-11 at 16 23 30" src="https://user-images.githubusercontent.com/3082655/61058878-541e4e80-a3f8-11e9-9d06-dc13174ca1b0.png">

#### What it does

- [x] close https://github.com/theia-ide/theia/issues/5058 - complete integration of view containers in the plugin system and shell
- [x] https://github.com/theia-ide/theia/issues/4199 - support `onView` activation event 
- [x] register views after all view containers are registered
- [x] fix https://github.com/theia-ide/theia/issues/4174 - support view/title menu contribution
- [x] support more view/title menu contribution
- [x] pre-defined views
  - [x] debug
  - [x] scm
  - [x] explorer
  - [x] test
    - [x] a CQ for copied icon - https://dev.eclipse.org/ipzilla/show_bug.cgi?id=20431
- [x] a command to toggle view container in the view menu and the command palette 
- [x] commands to toggle view containers contributed by plugin system
- [x] `Open View` command
- [x] quick open `view` prefix
- [x] implemented TreeItem2.label proposed API for sample extension
3ea5c9ddbebd8ec68e3b821f9c39c3ec785fde97/src/vs/vscode.proposed.d.ts#L1436-L1477
  - [x] a CQ for copied code - https://dev.eclipse.org/ipzilla/show_bug.cgi?id=20431
- [x] fix https://github.com/theia-ide/theia/issues/4637 - single-view view-container mode
  - [x] view caption should be inlined into the view container caption
  - [x] view toolbar items should be inlined into the view container toolbar
    - [x] scm
    - [x] explorer
    - [x] plugin views
  - [x] toolbar context menu should allow to hide/show view parts
- [x] show view container part's toolbar only on hover
- [x] plugin tree views: not enough space between icon and text if the width is small
- [x] restoration layout of view container on reload
  - [x] debug - no stateful
  - [x] scm - input message is preserved
  - [x] explorer
    - [x] fix https://github.com/theia-ide/theia/issues/1781 - restore internal state of the navigator
  - [x] test
  - [x] plugin views
    - [x] remove stale widget, i.e. which are not registered anymore
    - [x] move widgets to new view containers if package.json was changed
    - [x] restore internal state of plugin tree widgets
- [x] view container activation
  -  [x] should restore active part on restore of shell layout
  -  [x] should restore previous active part on activation, or activate the first expanded or visible part, or receive focus itself
-  [x] shell: handle activation and reveal requests recursively
- [x] fix UI tests
- [x] shell layout versioning and backward-compatibility
- [x] integrate https://github.com/theia-ide/theia/pull/5718

#### How to test

- install https://github.com/microsoft/vscode-extension-samples/tree/master/tree-view-sample from sources under `plugins folder
- run Theia and check that all views are added
- modify package.json to move views to other view containers (scm, debug, test) and test that views are added there as well

What to pay attention:
- keyboard navigation: one can navigate between view parts with keyboard plus collapse and expand them, try to drag and drop as well
- activation (focus): a proper part is focused
- restoration of layout
- explorer in multi-root workspace

#### Review checklist

- [x] as an author, I have thoroughly tested my changes and carefully reviewed following [the review guidelines](https://github.com/theia-ide/theia/blob/ak/pr_template/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/ak/pr_template/doc/pull-requests.md#reviewing)

akosyakov commented on packages/core/src/browser/view-container.ts:
```diff
@@ -126,22 +182,44 @@ export class ViewContainer extends BaseWidget implements StatefulWidget, Applica
+        const toDisposeOnHide = new DisposableCollection();
+        event.items.forEach((item, index) => {
+            const commandId = `__command_${this.id}_${index}`;
+            this.commandRegistry.registerCommand({ id: commandId }, {
```
`toDisposeOnHide.push(`

akosyakov commented on packages/plugin-ext/src/main/browser/plugin-contribution-handler.ts:
```diff
@@ -136,11 +136,17 @@ export class PluginContributionHandler {
         if (contributions.viewsContainers) {
```
register commands and menu items before other contributions in order to make sure that they are present when view containers are created

akosyakov commented:
@tetchel @ariel-bentu could you give it a try with your extensions contributing view containers and views? Feedback is very welcomed!

tetchel commented:
It does not appear to have worked for me.

```
        "views": {
            "explorer": [
                {
                    "id": "%viewID%",
                    "name": "%viewName%"
                }
            ]
        },
```
where:
```
    "viewName": "Codewind",
    "viewID": "ext.cw.explorer",
```

this is the same snippet that works in VS Code.

VS Code: 
![image](https://user-images.githubusercontent.com/8888438/61384759-3a10c000-a87f-11e9-834b-60fe40fd9d90.png)

Theia:
![image](https://user-images.githubusercontent.com/8888438/61384796-4dbc2680-a87f-11e9-8b53-77a9ff6e1db4.png)

akosyakov commented:
@tetchel why do you use internalization for internal constants? it should be used for user visible text. Anyway i will add handling for it, since it appears to happen :)

akosyakov commented:
@tetchel Could you also try to do `Reset Workbench Layout`? Since layout changed so much old layout data probably cannot be loaded. @svenefftinge @spoenemann @kittaakos I'm thinking to add a version to layout data, so we increase it each time to reject old layout state.

tetchel commented:
Well the viewName is user-visible, I suppose it's true that there's no point in externalizing the viewID. let me see if that fixes it.

kittaakos commented:
> I'm thinking to add a
> layout version to layout data

I had a deja vu after reading your comment, @akosyakov 😊 We already have something for it:
https://github.com/theia-ide/theia/blob/613069fc87169738999adab4044bc0530e54cb74/packages/core/src/browser/shell/application-shell.ts#L48 

The related task was: https://github.com/theia-ide/theia/issues/1075

akosyakov commented:
@kittaakos oh, cool, i can just increase it!

akosyakov commented:
@tetchel could you pull and try again, I've increased layout data, so it should be loaded from scratch for you now

tetchel commented:
After changing the `viewID` and pulling, it worked! 🎉 
![image](https://user-images.githubusercontent.com/8888438/61396802-83b8d500-a896-11e9-9190-96c807192424.png)

tetchel commented:
Do you mean the differences in text on the `Projects` and `No Projects` item? That's in my code. or is there some other difference?

akosyakov commented:
@tetchel there is for example `Click here to create...` in VS Code, but not in Theia

tetchel commented:
Yea that is intentional because in vs code clicking that item runs the Create Project `command`. But I removed it in theia because on-click tree item commands don't work

akosyakov commented:
ok, do we have an issue already for ` But I removed it in theia because on-click tree item commands don't work.` @tetchel ?

tetchel commented:
I thought we did, but it was just a few comments on #5661. I just opened https://github.com/theia-ide/theia/issues/5744

vince-fugnitto commented:
I'll add comments as I test the PR:
- The are some issues with the multi-root support in the `explorer`. 
It's much harder to get the `Add Folder to Workspace` context menu item to appear.
When adding a root to the workspace the explorer is not updated until the application is reloaded.

kittaakos commented:
Remarks during the verification: (I am going to update this post.)
 - [minor][styling]: Incorrectly styled border in the view container part header. The border has a tinny top annd left offset.
![Screen Shot 2019-07-19 at 08 23 44](https://user-images.githubusercontent.com/1405703/61513909-d90ff600-a9fe-11e9-81b3-8d300984b5ad.jpg)

akosyakov commented:
> I would have expected, the layout resets automatically when the layout data versions are incompatible. Showing this as a notification would be awesome.

Was not it reset? I guess an error comes from the original implementation, should we remove it?

akosyakov commented:
@lmcbout @vince-fugnitto I've removed a visible predicate for `Add Workspace Folder` command, so it is always available in the command palette. I think it was bogus even before since there should be a way to add a folder to the empty workspace and before command was not available without at least one folder.

<img width="1792" alt="Screen Shot 2019-07-20 at 17 39 54" src="https://user-images.githubusercontent.com/3082655/61580808-259a2500-ab16-11e9-8d2e-d916862c9af8.png">

I've added a delegating command to the navigator and this one has the visibility predicate to show it only on roots. If no workspace is opened a user can get access to a root folder context menu via right clicking on the space below all nodes. It is aligned with VS Code.

<img width="1792" alt="Screen Shot 2019-07-20 at 17 39 50" src="https://user-images.githubusercontent.com/3082655/61580809-259a2500-ab16-11e9-8f51-df18d040720f.png">

akosyakov commented:
@kittaakos I cannot get any top and left offset  in the view container part header:
<img width="607" alt="Screen Shot 2019-07-20 at 17 55 05" src="https://user-images.githubusercontent.com/3082655/61580918-b58c9e80-ab17-11e9-8e45-1782c22ea4a0.png">

Maybe i don't understand what you mean.

akosyakov commented:
@kittaakos > [key-navigation]: It does not work reliably.

Do you mean that it should be possible to navigate between headers with keys? One can move focus between headers with tab and shift+tab. It is the same in VS Code.

akosyakov commented:
>  [bug] View container part header loses the focus after DND.

That's a good one! I was relying on the shell current widget events to detect focus of parts. It is bogus since parts are actually not tracked. I'm going to change it to rely on `focus` event coming from part's node or its child nodes.

akosyakov commented:
> [bug] Open Debug > Close Debug > Reopen Debug > Right click on view container part header:

@kittaakos a good one as well, it is turned out that implementation of `unregisterCommand` was bogus, it was not align with `register` in the case if a handler is passed and as consequences leaked a handler and a referenced widget from it. A fix is here: https://github.com/theia-ide/theia/pull/5665/commits/e507b9adcbce1f2373194928d73921b9535883d7#diff-ab6e2d4c57658cf60e550cfe9ba87eac

akosyakov commented:
> [minor]: If a view container part is not hidable, do not show its context menu item:

@kittaakos it is similar in VS Code, see `Explorer` for example. I don't have a strong opinion on it. We can always hide them or show as VS Code does. Maybe other have some reasons for one or another way. @vince-fugnitto @lmcbout ?

akosyakov commented:
> [critical] [tree-view-example]: Open the example, hide a few view container parts, refresh your browser. The layout is correctly restored but I can see this in the browser, and I had to force quit from my browser:

@kittaakos I cannot reproduce it, could you try again? Maybe it was fixed by latest changes. If not we need to talk 😄

akosyakov commented:
> [tree-view-example]: I can see these when I start my instance although it seems to work:

@kittaakos I could reproduce it on master. The issue was that ftp tree throws an error but as a string and it gets swallowed during going via RPC. Not it should be wrapped in the error object and presented to a user. Please pay attention that you should not see it before a user make ftp tree visible. It should be considered for the initial layout as well as for restored. Relevant commit: https://github.com/theia-ide/theia/pull/5665/commits/392267aca382fe805d9b0714adc65baa18482b18

kittaakos commented:
I am verifying your changes once more, @akosyakov.
Please note the red CIs.

kittaakos commented:
> it is similar in VS Code, see `Explorer` for example

Let's stick to your solution.

akosyakov commented:
Fixed tests, again dependencies to monaco leaked to spec.ts. Cannot wait when we can just run unit tests within the app.

kittaakos commented:
> I cannot reproduce it, could you try again?

I still see the same error. I have tried to wipe the `~/.theia` conten, the error still persists.
I did the followings:
 - Start Theia.
 - Start the `./vscode-extension-samples/tree-view-sample/` as an instance.
 - Hide one of the view container parts in the running in instance.
 - Refresh the browser window of the hosted instance.

```
tree.ts:15 Uncaught (in promise) RangeError: Maximum call stack size exceeded
    at PluginTree.../../packages/core/lib/browser/tree/tree.js.TreeImpl.getRootNode (tree.ts:290)
    at PluginTree.../../packages/core/lib/browser/tree/tree.js.TreeImpl.getRootNode (tree.ts:294)
    at PluginTree.../../packages/core/lib/browser/tree/tree.js.TreeImpl.getRootNode (tree.ts:294)
    at PluginTree.../../packages/core/lib/browser/tree/tree.js.TreeImpl.getRootNode (tree.ts:294)
    at PluginTree.../../packages/core/lib/browser/tree/tree.js.TreeImpl.getRootNode (tree.ts:294)
    at PluginTree.../../packages/core/lib/browser/tree/tree.js.TreeImpl.getRootNode (tree.ts:294)
    at PluginTree.../../packages/core/lib/browser/tree/tree.js.TreeImpl.getRootNode (tree.ts:294)
    at PluginTree.../../packages/core/lib/browser/tree/tree.js.TreeImpl.getRootNode (tree.ts:294)
    at PluginTree.../../packages/core/lib/browser/tree/tree.js.TreeImpl.getRootNode (tree.ts:294)
    at PluginTree.../../packages/core/lib/browser/tree/tree.js.TreeImpl.getRootNode (tree.ts:294)
```
![Screen Shot 2019-07-22 at 09 17 09](https://user-images.githubusercontent.com/1405703/61613375-7cf8dc00-ac61-11e9-9f78-c7ae2eec3a01.jpg)

kittaakos commented:
> Maybe i don't understand what you mean.

![Screen Shot 2019-07-22 at 09 20 11](https://user-images.githubusercontent.com/1405703/61613603-0dcfb780-ac62-11e9-8108-f92d36fd9602.jpg)

kittaakos commented:
> headers with tab and shift+tab. It is the same in VS Code.

Nice, it works much better when I use the desired kaybinding :)

It seems, when a part is hidden, we still try to focus it. Hide a view container part, and press `tab` and `shift`+`tab`:

![screencast 2019-07-22 09-23-12](https://user-images.githubusercontent.com/1405703/61613964-e62d1f00-ac62-11e9-8c54-a299bcc4c2cb.gif)
I have to press `tab` twice to get from the `Call Stack` to `Variables` same with the other direction with `shift`+`tab`.

vince-fugnitto commented:
> @lmcbout @vince-fugnitto I've removed a visible predicate for `Add Workspace Folder` command, so it is always available in the command palette. I think it was bogus even before since there should be a way to add a folder to the empty workspace and before command was not available without at least one folder.
> 
> <img alt="Screen Shot 2019-07-20 at 17 39 54" width="1792" src="https://user-images.githubusercontent.com/3082655/61580808-259a2500-ab16-11e9-8d2e-d916862c9af8.png">
> 
> I've added a delegating command to the navigator and this one has the visibility predicate to show it only on roots. If no workspace is opened a user can get access to a root folder context menu via right clicking on the space below all nodes. It is aligned with VS Code.
> 
> <img alt="Screen Shot 2019-07-20 at 17 39 50" width="1792" src="https://user-images.githubusercontent.com/3082655/61580809-259a2500-ab16-11e9-8f51-df18d040720f.png">

It all sounds good to me! :)
In the future we can think about having the `add folder to workspace` directly from the `file` menu (#5522)

vince-fugnitto commented:
@akosyakov more of a question, should the `explorer` display this context menu when right clicking the `explorer` label? I assume this is meant more for when you have multiple different sections of a view that you wish to show/hide.

<div align='center'>

![image](https://user-images.githubusercontent.com/40359487/61630710-6bf69d80-ac56-11e9-9f4d-dc8d22deeabe.png)

</div>

lmcbout commented:
@akosyakov When testing this PR, here what happen on my PC:
OS: Ubuntu 16.04
Theia:0.8.0 with your patch ( commit 629c3a099f2 )
The error on server side: 'maxResults',  spawn ENOMEM,   Connection got disposed.

My Test:

1- Start BE-FE
--------------------
Observation: Start a frontend -> Open a workspace, Open context menu, expand a view (Outline) then it goes OFFLINE
Note: the"Add folder to workspace" is available only on the first system load, never after, don't know if it could be related or not 


Server console log after a few seconds:

root INFO [nsfw-watcher: 2882] Started watching: /home/lmcbout/theia/vinceTest
Started watching: /home/lmcbout/theia/vinceTest
root ERROR TypeError: Cannot read property 'maxResults' of undefined
    at SearchInWorkspaceWidget.push.../../packages/search-in-workspace/lib/browser/search-in-workspace-widget.js.SearchInWorkspaceWidget.renderNotification (http://localhost:3000/88.bundle.js:2356:112)
    at SearchInWorkspaceWidget.push.../../packages/search-in-workspace/lib/browser/search-in-workspace-widget.js.SearchInWorkspaceWidget.renderSearchField (http://localhost:3000/88.bundle.js:2380:33)
    at SearchInWorkspaceWidget.push.../../packages/search-in-workspace/lib/browser/search-in-workspace-widget.js.SearchInWorkspaceWidget.renderSearchAndReplace (http://localhost:3000/88.bundle.js:2330:32)
    
root ERROR Error: spawn ENOMEM
    at ChildProcess.spawn (internal/child_process.js:366:11)
    at spawn (child_process.js:538:9)
    at Object.fork (child_process.js:113:10)
    at IPCConnectionProvider.fork (/home/lmcbout/theia/validation/theia/packages/core/lib/node/messaging/ipc-connection-provider.js:103:31)

root ERROR Error: Connection got disposed.
    at Object.dispose (http://localhost:3000/bundle.js:83245:25)
    at http://localhost:3000/bundle.js:85058:56
    at CallbackList.../../node_modules/vscode-jsonrpc/lib/events.js.CallbackList.invoke (http://localhost:3000/bundle.js:81988:39)


2- RELOAD the BE-FE
--------------------

 Also, Reload a second time the FE 
Server console -> root INFO [hosted-plugin: 4686] PLUGIN_HOST(4686) starting instance
root ERROR TypeError: Cannot read property 'maxResults' of undefined

root ERROR Uncaught Exception:  Error: Connection got disposed.


3- RELOAD the BE-FE a third time:
------------------------------
Server console:
root INFO [hosted-plugin: 5634] PLUGIN_HOST(5634) starting instance
root ERROR TypeError: Cannot read property 'maxResults' of undefined

root ERROR Error: spawn ENOMEM

FE: did not start 
-----------------

kittaakos commented:
@lmcbout, can you try to reset the workbench layout? Does the problem persist?

vince-fugnitto commented:
> @lmcbout, can you try to reset the workbench layout? Does the problem persist?

I don't see the same issues as @lmcbout.
I'm able to always see the menu item for `Add Folder to Workspace...` (even after subsequent reloads), and I don't get errors regarding `maxResults`.

lmcbout commented:
@akosyakov  I probably had an old workspace reload with mismatch BE-FE at some point that trigger the issue reported just before. I started the BE-FE on a brand new port and I did not reproduce the error, so just disregard what I put previously.

lmcbout commented:
>[minor]: If a view container part is not hidable, do not show its context menu item:

I don't have a strong opinion here either, I guest we should go like VSCode

akosyakov commented:
@kittaakos regarding https://github.com/theia-ide/theia/pull/5665#issuecomment-513672124

The space is above it not a header border, but another element: a split panel resize handle. The header btw does not use border focusing, but outline. Otherwise it causes flickering on focus.

akosyakov commented:
> RangeError: Maximum call stack size exceeded

@kittaakos I was able to reproduce it. It caused by a race condition between restoring the tree widget state for `Test View` and synching it with the running plugin. You don't need to hide it but expand several levels of children. I've fixed it by awaiting complete refresh from the previous state: https://github.com/theia-ide/theia/pull/5665/commits/624192b7fc22d5a295e92a639da9a3196d9e60c5

I've also had to fix `TreeModel.refresh` method. It was not awaiting refresh of all expanded children as documented, now it should: https://github.com/theia-ide/theia/pull/5665/commits/5f6522045706adb7b48741d1706e92aa366388e1 Unfortunately test file was bogusly formatted before 🙈 You will need to enable `hide whitespace changes` to review it.

kittaakos commented:
I am verifying the changes once more.

kittaakos commented:
> but another element: a split panel resize handle

I see. 😞

kittaakos commented:
`tab`/`shift`+`tab` traversal is still odd; I can focus the scroll-bar, for instance:
![screencast 2019-07-23 10-54-50](https://user-images.githubusercontent.com/1405703/61697908-830dbc00-ad38-11e9-8585-a767f0ca8827.gif)

kittaakos commented:
`Open View`
![Screen Shot 2019-07-23 at 10 57 14](https://user-images.githubusercontent.com/1405703/61698034-b8b2a500-ad38-11e9-8103-476eedef0b46.jpg)
 does not yet follow the pattern we use, please add the ellipses (`...`)

kittaakos commented:
After DNDing the view container parts, I ended up with this broken tree. No errors in the log:
![Screen Shot 2019-07-23 at 10 59 00](https://user-images.githubusercontent.com/1405703/61698196-fd3e4080-ad38-11e9-8a94-894011b7f4f9.jpg)

kittaakos commented:
View state incorrectly restored.
 - Open `Explorer` and make sure the `Tree-View-Sample` is expanded, others are collapsed.
 - Open another view,  I hid one of the parts.
 - Refresh the browser.
 - Activate the `Explorer`, `Tree-View-Sample` is collapsed. Should be expanded instead.

![screencast 2019-07-23 11-04-49](https://user-images.githubusercontent.com/1405703/61698956-480c8800-ad3a-11e9-97a5-b26173ca2c50.gif)

kittaakos commented:
I do not know if it is related, I ended up having multiple selections in the outline. I noted tree-related changes in the PR.

![screencast 2019-07-23 11-17-18](https://user-images.githubusercontent.com/1405703/61699757-c9b0e580-ad3b-11e9-93e8-a78b127c73d5.gif)

kittaakos commented:
> I do not know if it is related

I can reproduce it with the `Outline`.

kittaakos commented:
> I ended up with this broken tree.

Try to DND the `File Explorer`, it happens quite often:

![Screen Shot 2019-07-23 at 11 27 09](https://user-images.githubusercontent.com/1405703/61700379-e7327f00-ad3c-11e9-993c-f53e04daee80.jpg)

akosyakov commented:
@kittaakos 

https://github.com/theia-ide/theia/pull/5665#issuecomment-514119617 should be fixed

I could not reproduce https://github.com/theia-ide/theia/pull/5665#issuecomment-514127645, although I changed tree impl again to fix `refresh` with less changes. Maybe it was causing the issue. Could you try again?

And remaining issues are hard:
- https://github.com/theia-ide/theia/pull/5665#issuecomment-514130527 - perfect scrollbar is not capable to keep scroll after reattaching host node to another node. Not sure what can be done about it. It is kind of a glitch but a user can recover with scrolling again.
- https://github.com/theia-ide/theia/pull/5665#issuecomment-514123844 - storeState is called for hidden widget, as consequence available size and size of parts are zero, we need to preserver sizes between attachment/detachment to restore them properly.
- https://github.com/theia-ide/theia/pull/5665#issuecomment-514119184 - (1) there is no way to disable focusing of scrolls, it is enabled for accessibility (https://github.com/mdbootstrap/perfect-scrollbar/issues/425). I think we just should keep it for now, maybe open an issue in perfect scrollbar to make it configurable. (2) We disabled outline based highlighting for trees on purpose, but it does not work nicely with empty nodes. I could add a special css class for empty tree.

kittaakos commented:
@akosyakov, I have tried it again, on Windows this time... It worked. I did not spend too much time on this verification cycle though. Perhaps someone else can review your work.

One remark on the whole PR, I think the `tab`/`shift`+`tab` navigation is not very useful, can we remove it? Or do you consider it as a nice to have feature?

By the way, I could reproduce the multi-selection issue in the `Outline` on Windows as well. But not consistently.

akosyakov commented:
> I have tried it again, on Windows this time... It worked. I did not spend too much time on this verification cycle though. Perhaps someone else can review your work.

Yes, at least changes in core package. There are important changes to allow storing layout of composite widgets and activating them properly on restoring. cc @svenefftinge @spoenemann 

> One remark on the whole PR, I think the tab/shift+tab navigation is not very useful, can we remove it? Or do you consider it as a nice to have feature?

It is nice to have, but if we cannot make it right. I would revert it and file an issue to resolve it later. It is not so important.

> By the way, I could reproduce the multi-selection issue in the Outline on Windows as well. But not consistently.

Could you elaborate how? I could not get from previous comments. Is it somehow related to restoring layout. Now selected nodes are stored as well to restore selection. Maybe you can spot something suspicious here: https://github.com/theia-ide/theia/pull/5665/files#diff-bc06031588a7b793bba8457dd8ee55ba

akosyakov commented:
I would try to reiterate on https://github.com/theia-ide/theia/pull/5665#issuecomment-514393891 today again and if i cannot get something working would try to remove it from the scope.

kittaakos commented:
> It is nice to have, but if we cannot make it right. I would revert it and file an issue to resolve it later. It is not so important.

That's what I thought as well. But it's your call.

> Could you elaborate how?

I do not know the steps to reproduce it. Open the Outline and it happens from time to time: https://github.com/theia-ide/theia/pull/5665#issuecomment-514127645

akosyakov commented:
@kittaakos I've managed to address remaining issues from https://github.com/theia-ide/theia/pull/5665#issuecomment-514393891 Could you try again please?

- scroll bar focus is disabled now, it is up to a widget to decide what to focus and how to handle it
- empty tree widget is highlighted now
- scroll bar position is preserved now between detaching and attaching of a widget
- view container state is preserved now between collapsing and expanding them, if a reload happens when a container is collapsed then a preserved state is stored as a part of a layout information

kittaakos commented:
> Could you try again please?

I am reviewing it now.

kittaakos commented:
I see this logged into the browser console when I change the focused view on the left side-bar. Is this expected or leftover?

![Screen Shot 2019-07-31 at 08 25 20](https://user-images.githubusercontent.com/1405703/62188482-ddc49a80-b36c-11e9-8980-9a423f60533d.jpg)

kittaakos commented:
After resetting the workbnech layout (in my `tree-view-sample` hosted instance), from time to time I end up with a collapsed view state. I cannot reproduce this consistently.

![Screen Shot 2019-07-31 at 08 28 08](https://user-images.githubusercontent.com/1405703/62188691-6e02df80-b36d-11e9-9ae1-c0372d925b56.jpg)

kittaakos commented:
> I cannot reproduce this consistently.

The image, I have [attached above](https://github.com/theia-ide/theia/pull/5665#issuecomment-516716506), shows that I have `Theia` as the opened workspace, but it should be the `tree-view-sample`. Probably a bug in the plug-ins.

![Screen Shot 2019-07-31 at 08 28 08](https://user-images.githubusercontent.com/1405703/62189006-3a748500-b36e-11e9-922a-48f458c29eb4.jpg)

kittaakos commented:
Very nice you added the border when the part has the focus but it has no content 👍

![Screen Shot 2019-07-31 at 08 40 54](https://user-images.githubusercontent.com/1405703/62189355-0483d080-b36f-11e9-9653-4af29e943162.jpg)


[later]: Perhaps we should align the debug view with other VS Code contributions and get rid of the `Not running` from the view. The users can see it is not running and they already got used to `Debug` view from VS Code.

![Screen Shot 2019-07-31 at 08 41 05](https://user-images.githubusercontent.com/1405703/62189426-37c65f80-b36f-11e9-907f-ffd6d7118f46.jpg)

kittaakos commented:
Start debug, have a breakpint, refresh the browser, start another debug session:
```
Uncaught Error: the debug session connection is disposed, id: 0c664f7b-10d2-45b8-9a81-259fbce6c807
    at DebugSessionConnection.push.../../packages/debug/lib/browser/debug-session-connection.js.DebugSessionConnection.checkDisposed (debug-session-connection.ts:129)
    at DebugSessionConnection.push.../../packages/debug/lib/browser/debug-session-connection.js.DebugSessionConnection.newEmitter (debug-session-connection.ts:284)
    at DebugSessionConnection.push.../../packages/debug/lib/browser/debug-session-connection.js.DebugSessionConnection.getEmitter (debug-session-connection.ts:278)
    at DebugSessionConnection.push.../../packages/debug/lib/browser/debug-session-connection.js.DebugSessionConnection.doFire (debug-session-connection.ts:275)
    at DebugSessionConnection.push.../../packages/debug/lib/browser/debug-session-connection.js.DebugSessionConnection.handleEvent (debug-session-connection.ts:258)
    at DebugSessionConnection.push.../../packages/debug/lib/browser/debug-session-connection.js.DebugSessionConnection.handleMessage (debug-session-connection.ts:214)
    at WebSocketChannel.fireMessage (debug-session-connection.ts:146)
    at WebSocketChannel.../../packages/core/lib/common/messaging/web-socket-channel.js.WebSocketChannel.handleMessage (web-socket-channel.ts:49)
    at WebSocketConnectionProvider.socket.onmessage (ws-connection-provider.ts:74)
```

kittaakos commented:
When I open a view container part from the _Command Palette_ (`Open View...`), it does not accept focus. The opened viewes work, but I see the warnings in the console.

![Screen Shot 2019-07-31 at 08 49 32](https://user-images.githubusercontent.com/1405703/62189758-203ba680-b370-11e9-994c-5c5bd483053a.jpg)

kittaakos commented:
> Is this expected or leftover?

Either it is an intentional logging or not, it shows, my view does not have a `title`. Weird. The view works by the way.

![Screen Shot 2019-07-31 at 08 57 08](https://user-images.githubusercontent.com/1405703/62190350-6c3b1b00-b371-11e9-987d-5b0a8d5d2d1a.jpg)

kittaakos commented:
I have verified the changes described [here](https://github.com/theia-ide/theia/pull/5665#issuecomment-516165853), very nice work!

akosyakov commented:
> Either it is intentional logging or not, it shows, my view does not have a title. Weird. The view works by the way.

Not it is not intentional, i remove it. Thanks @kittaakos

akosyakov commented:
> Start debug, have a breakpint, refresh the browser, start another debug session:

It's a known issue: https://github.com/theia-ide/theia/issues/5504

akosyakov commented:
> [later]: Perhaps we should align the debug view with other VS Code contributions and get rid of the Not running from the view. The users can see it is not running and they already got used to Debug view from VS Code.

Originally we wanted to align it with Chrome dev tools.

akosyakov commented:
> When I open a view container part from the Command Palette (Open View...), it does not accept focus. The opened viewes work, but I see the warnings in the console.

The check is not reliable :( It just wait one animation frame, it is obviously is not always enough.

akosyakov commented:
@kittaakos I've pushed a commit to fix https://github.com/theia-ide/theia/issues/1707 Now activation check should be performed only when a widget really received activation request and based on focus events not timing.

akosyakov commented:
@vince-fugnitto @lmcbout @kittaakos @azatsarynnyy @AlexTugarev We need to get it landed. It blocks improvements for advance debugging and progress notifications. Please test if you can not everythign but some bits or approve if are fine.

I had an offline review session with @svenefftinge yesterday on the core part. He is fine with changes.

AlexTugarev commented:
in vscode a view is not ther if the `when` predicate is falsely.

```
{
	"id": "jsonOutline",
	"name": "Json Outline",
	"when": "jsonOutlineEnabled"
}
```

![2019-08-01 10 33 03](https://user-images.githubusercontent.com/914497/62278196-e3db7980-b447-11e9-8377-155c50dfef82.gif)

but that's not a big deal I think.

AlexTugarev commented:
<img width="301" alt="Screen Shot 2019-08-01 at 10 31 45" src="https://user-images.githubusercontent.com/914497/62278424-54829600-b448-11e9-8205-a924a609bc97.png">

which part is responsible to decorate the tree nodes?

AlexTugarev commented:
Take the `JSON OUTLINE`: arrow keys are working nicely! but `enter` does not reveal the element in the editor.

AlexTugarev commented:
<img width="316" alt="Screen Shot 2019-08-01 at 10 48 21" src="https://user-images.githubusercontent.com/914497/62279194-e7700000-b449-11e9-8cb4-f035aff11635.png">

can we increase the min height of views? it's ~120px in vscode, but only ~50px here ;-)

AlexTugarev commented:
1. tried with default layout from the mentioned example extension.
2. then I changed the layout and restarted Theia
3. hard reload didn't help to make it effective 
4. I had to `reset the workbench layout` to make it work

AlexTugarev commented:
```
root WARN Widget was activated, but did not accept focus: explorer-view-container
root WARN Widget was activated, but did not accept focus: scm-view-container
root WARN Widget was activated, but did not accept focus: scm-view-container
root WARN Widget was activated, but did not accept focus: plugin-view:jsonOutline
root WARN Widget was activated, but did not accept focus: outline-view
root WARN Widget was activated, but did not accept focus: scm-view-container
```
is logged often, is this ok?

akosyakov commented:
> which part is responsible to decorate the tree nodes?

> Take the JSON OUTLINE: arrow keys are working nicely! but enter does not reveal the element in the editor.

@AlexTugarev please file follow-up issues to improve plugin tree widgets, it is not related to view containers and view generally

akosyakov commented:
> in vscode a view is not ther if the when predicate is falsely.

@AlexTugarev It would be worth a follow-up issue as well.

AlexTugarev commented:
I tried to break keyboard navigation, but failed to do so ;-)

the only thing which feels _strange_ is the highlighting of a selected item per view. look at this screenshot and tell me which node will be expanded on <kbd>-></kbd>

<img width="351" alt="Screen Shot 2019-08-01 at 11 07 24" src="https://user-images.githubusercontent.com/914497/62280608-901f5f00-b44c-11e9-8e01-9810f853bb10.png">

akosyakov commented:
> the only thing which feels strange is the highlighting of a selected item per view. look at this screenshot and tell me which node will be expanded on ->

nothing seems to be selected

AlexTugarev commented on packages/scm/src/browser/scm-widget.tsx:
```diff
@@ -100,16 +94,8 @@ export class ScmWidget extends ReactWidget implements StatefulWidget {
-        } else if (area === 'right') {
-            this.shell.rightPanelHandler.refresh();
-        }
+        this.title.label = repository ? repository.provider.label : 'no repository opened';
```
![Screen Shot 2019-08-01 at 15 25 43](https://user-images.githubusercontent.com/914497/62297082-cbcc2000-b470-11e9-9644-57431abe53ce.png)

cannot see the repository opened anymore.

![Screen Shot 2019-08-01 at 15 27 28](https://user-images.githubusercontent.com/914497/62297111-e3a3a400-b470-11e9-9ead-3e4402ab3a85.png)

this is in vscode: `B` is the repo name on FS ;-)

akosyakov replied:
Our scm widget behaves different to scm widget from VS Code for multiple repos:
- in VS Code you have a new part for each repo, so they have to disambiguate them with a label
- in Theia there is always only one part, and a user see a repo and can change in the status bar

I would say you should file an issue if you want to align it with VS Code, although i don't see how their approach is better? It introduces custom handling for a widget instead of relying on the same code.

AlexTugarev replied:
Thanks for explaining!

> in Theia there is always only one part, and a user see a repo and can change in the status bar

I like the fact that it's in the status bar a lot, and coercing multiple instances of the scm widget into that area could be less useful.

AlexTugarev commented:
>>  the only thing which feels strange is the highlighting of a selected item per view. look at this screenshot and tell me which node will be expanded on ->

> nothing seems to be selected

the folder named `dev-packages` will be expanded ;-) 
that's the surprising/confusing bit of it.

akosyakov commented:
@AlexTugarev I've changed activation check again to fix https://github.com/theia-ide/theia/pull/5665#issuecomment-517192583

Now it waits only for last activated widget, i.e. if a user clicks fast on different side bars then check is cancelled for previous widgets. It waits 2s and after that reports the issue. I've tested it by commenting out `onActivatioRequest` method for the debug widget.

akosyakov commented:
@AlexTugarev https://github.com/theia-ide/theia/pull/5665#issuecomment-517192097 should be fixed as well, now a plugin view is moved to a proper view container when view container is created

Please have a look again. Please also track all other issues for current widgets which you discovered we can handle them with smaller follow-up PRs.

akosyakov commented:
@AlexTugarev I've figured out why view container layout state is not preserver on reload always. There were a case when a state is restored on first reload, but since the view is never opened, it is lost on second reload. Now on first reload the last visible state is set to restored one, see first line in `ViewContainer.restoreState`.

AlexTugarev commented:
https://github.com/theia-ide/theia/pull/5665#issuecomment-517192097 is solved only partially. I do see some changes after restart, i.e. a new view handle, but it seem to be not properly initialized: there is no title and it doesn't appear in the context menu.

after a hard reset of the layout it's there as it should be.
 
![2019-08-02 13 48 15](https://user-images.githubusercontent.com/914497/62368247-59207a80-b52c-11e9-83e3-69ea9cd35a68.gif)

akosyakov commented:
@AlexTugarev https://github.com/theia-ide/theia/pull/5665#issuecomment-517672057 was the hard one. When view container part is disposed then the wrapped widget should be disposed (closed) as well and other way around. But there is also a case when a wrapped widget is detached, not disposed, in this case only a part should be disposed, not a wrapped widget (edited). It was not handled. Could you try again?

AlexTugarev commented:
![Screen Shot 2019-08-02 at 15 44 10](https://user-images.githubusercontent.com/914497/62374553-7f4e1680-b53c-11e9-8333-4bbbd7eb623f.png)

verified twice: it still happens.

akosyakov commented:
@AlexTugarev triy again, I've messed up rebase last time

AlexTugarev commented:
Verified, the most [recent change](https://github.com/theia-ide/theia/compare/94c8539d30e2ddcf6dfb0d053c08b601924acaa4..80c4e8651979abbcd86eab68406d64cb608139b8) does solve the issue with moved views. Cool!

akosyakov commented:
I'm merging if noone objects when the build is green. Please open new issues if something breaks afterwards. It would be my top-priority to fix them.
